plugins/inputs/azure_monitor/azure_monitor.go

```diff
@@ -0,0 +1,856 @@
+		if len(target.Metrics) > 0 {
+			continue
+		}
+
+		am.Log.Info("Getting metrics for target #", index+1)
+		waitGroup.Add(1)
+
+		go func(target *ResourceTarget) {
+			defer waitGroup.Done()
+
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			err := am.getResourceTargetMetrics(target)
+
+			if err != nil {
+				select {
+				case errChan <- err:
+				default:
+				}
+
+				cancel()
+				return
+			}
+		}(target)
+	}
+
+	waitGroup.Wait()
+
+	select {
+	case err := <-errChan:
+		return err
+	default:
+	}
+
+	return nil
+}
+
+func (am *AzureMonitor) getResourceTargetMetrics(target *ResourceTarget) error {
+	apiURL := am.buildMetricDefinitionsAPIURL(target)
+	body, err := am.getTargetResponseBody(apiURL)
+
+	if err != nil {
+		return err
+	}
+
+	err = target.setResourceTargetMetrics(body)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (am *AzureMonitor) checkTargetsMaximumMetrics() {
+	for _, target := range am.ResourceTargets {
+		if len(target.Metrics) <= maxMetricsPerRequest {
+			continue
+		}
+
+		for start := maxMetricsPerRequest; start < len(target.Metrics); start += maxMetricsPerRequest {
+			end := start + maxMetricsPerRequest
+
+			if end > len(target.Metrics) {
+				end = len(target.Metrics)
+			}
+
+			newTargetMetrics := target.Metrics[start:end]
+
+			var newTargetAggregation []string
+
+			for _, aggregation := range target.Aggregation {
+				newTargetAggregation = append(newTargetAggregation, aggregation)
+			}
+
+			newTarget := NewResourceTarget(target.ResourceID, newTargetMetrics, newTargetAggregation)
+
+			am.ResourceTargets = append(am.ResourceTargets, newTarget)
+		}
+
+		target.Metrics = target.Metrics[:maxMetricsPerRequest]
+	}
+}
+
+func (am *AzureMonitor) setResourceTargetsAggregation() {
+	for _, target := range am.ResourceTargets {
+		if len(target.Aggregation) > 0 {
+			continue
+		}
+
+		target.Aggregation = append(target.Aggregation, "Total", "Count", "Average", "Minimum", "Maximum")
+	}
+}
+
+func (am *AzureMonitor) Gather(acc telegraf.Accumulator) error {
+	err := am.refreshAccessToken()
+
+	if err != nil {
+		return err
+	}
+
+	err = am.collectResourceTargetsMetrics(acc)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func NewResourceTarget(
+	resourceID string,
+	metrics []string,
+	aggregation []string,
+) *ResourceTarget {
+	return &ResourceTarget{
+		ResourceID:  resourceID,
+		Metrics:     metrics,
+		Aggregation: aggregation,
+	}
+}
+
+func NewResourceGroupTarget(resourceGroup string, resources []*Resource) *ResourceGroupTarget {
+	return &ResourceGroupTarget{
+		ResourceGroup: resourceGroup,
+		Resources:     resources,
+	}
+}
+
+func NewAzureClient() *AzureClient {
+	return &AzureClient{
+		client:               &http.Client{},
+		accessToken:          "",
+		accessTokenExpiresOn: time.Time{},
+	}
+}
+
+func NewMetric() *Metric {
+	return &Metric{
+		name:   "",
+		fields: make(map[string]interface{}),
+		tags:   make(map[string]string),
+	}
+}
```

Comment:
I think we don't need to export those functions. And beyond I think you should rather setup the structs at the required place instead of wrapping it in those functions.